M1FExplained/Chapter25/Exercises/Exercise01.lean

```diff
@@ -0,0 +1,197 @@
+import Mathlib.Algebra.Group.Defs
+import Mathlib.Data.Complex.Basic
+import Mathlib.Analysis.Complex.Basic
+import Mathlib.Data.Nat.Basic
+abbrev complex_size_1 :=
+{ z : ℂ // ‖z‖  = 1 }
+
+instance one : One complex_size_1 := ⟨1,  by exact CstarRing.norm_one⟩
+
+@[simp]
+lemma coe_one : (↑(1 : complex_size_1) : ℂ) = 1 := by rfl
+
+lemma mul_closed (a b : complex_size_1) : ‖(a : ℂ) * b‖ = 1 := by
+  rw [norm_mul, a.2,b.2,mul_one]
+
+--- part i
+
+instance : Mul complex_size_1 where mul a b:= ⟨a.1 * b.1, mul_closed a b⟩
+
+@[simp]
+lemma coe_mul (a b : complex_size_1) : (↑(a * b) : ℂ) = ↑a * ↑b := by
+  rfl
+
+lemma ne_zero (a : complex_size_1) : (a : ℂ) ≠ 0 := by
+  intro h
+  apply_fun norm at h
+  rw [a.2, norm_zero,] at h
+  norm_num at h
+
+noncomputable instance : Group complex_size_1 where
+  mul a b:= ⟨a.1 * b.1, mul_closed a b⟩
+  mul_assoc := by 
+    intro a b c
+    ext1
+    simp [mul_assoc]
+  one := 1
+  one_mul := by
+    intro a
+    ext1
+    simp
+  mul_one := by
+    intro a
+    ext1
+    simp
+  inv z := ⟨z.1⁻¹, by rw [norm_inv,z.2,inv_one]⟩
+  mul_left_inv := by
+    intro a
+    ext1
+    simp [inv_mul_cancel]
+    rw [inv_mul_cancel]
+    exact ne_zero a
+
+--- part ii
+abbrev real_not_minus_1 :=
+{x : ℝ // x ≠ -1}
+
+instance zero : Zero real_not_minus_1 := ⟨0,  by norm_num⟩ 
+
+lemma add_closed2 (a b : real_not_minus_1) : (a : ℝ) * b + a + b ≠ -1 := by
+  by_contra h
+  have h₁ : (a.1 + 1) * (b.1 + 1) = 0 :=
+    calc
+      (a.1 + 1) * (b.1 + 1) = a.1 * b.1 + a.1 + b.1 + 1 := by ring
+      _ = 0 := by
+        rw [h]
+        norm_num
+  have h₂ : a.1 + 1 = 0 ∨ b.1 + 1 = 0 := Iff.mp zero_eq_mul (id (Eq.symm h₁))
```

Comment:
This proof looks like some complicated way to do something like `rw [zero_eq_mul] at h1`?

Reply:
Yeah you can just do `rw [mul_eq_zero] at h₁` and then change h2 to h1 on line 69